[CDAP-17786] Authorizer Error Support

dli357 opened the pull request:
Add native support for [`ensureAccess`](https://github.com/cdapio/cdap/blob/6dc6a3fa055c8f6e99da23f047d95c44b63a1ad6/cdap-security/src/main/java/io/cdap/cdap/security/authorization/AuthorizationUtil.java#L129) in AuthorizationEnforcer API and add support for custom error propagation from Authorizer in [`ensureOnePrivilege`](https://github.com/cdapio/cdap/blob/6dc6a3fa055c8f6e99da23f047d95c44b63a1ad6/cdap-security/src/main/java/io/cdap/cdap/security/authorization/AuthorizationUtil.java#L73).

Context: [CDAP-17786](https://cdap.atlassian.net/browse/CDAP-17786).

wyzhang commented on cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuthorizationEnforcer.java:
```diff
@@ -53,9 +54,38 @@
+   * @throws UnauthorizedException if the principal is not authorized to perform any of the actions on the entity
+   * @throws Exception if any other errors occurred while performing the authorization enforcement check
+   */
+  //void enforceAtLeastOneAction(EntityId entity, Principal principal, Set<Action> actions) throws Exception;
```
Is this needed?

dli357 replied:
Whoops, fixed

wyzhang commented on cdap-security/src/main/java/io/cdap/cdap/security/authorization/AuthorizationUtil.java:
```diff
@@ -74,17 +75,27 @@ public static void ensureOnePrivilege(io.cdap.cdap.proto.id.EntityId entityId, S
       } catch (UnauthorizedException e) {
-        // continue to next action
+        missingPermissions.addAll(e.getMissingPermissions());
+        if (!e.includePrincipal()) {
```
nit: includePrinciple = e.includePrincipal()

dli357 replied:
Prefer the original method, since this will use the last exception thrown instead of using it if any exception has that option.

wyzhang commented on cdap-security-spi/src/main/java/io/cdap/cdap/security/spi/authorization/AuthorizationEnforcer.java:
```diff
@@ -53,9 +54,38 @@
+   * @throws UnauthorizedException if the entity is not visible to the principal
+   * @throws Exception if any errors occurred while performing the check
+   */
+  default void isVisible(EntityId entityId, Principal principal) throws Exception {
```
btw, any reason we need a default implementation here?

dli357 replied:
We need it to be default for backwards compatibility. Additionally, we need it in the first place due to being unable to pass UnauthorizedExceptions consistently through the other isVisible method.

dli357 committed:
[CDAP-17786] Add native support for ensureAccess in AuthorizationEnforcer API and add support for custom error propagation from Authorizer in checkOnePrivilege

Fix testDataset to not depend on UnauthorizedException message

Update copyright year
